xcpmd/src/screen-module.c

```diff
@@ -0,0 +1,311 @@
+
+    if (new_backlight > 100) {
+        new_backlight = 100;
+    }
+    else if (new_backlight < old_backlight) {
+        new_backlight = old_backlight;
+    }
+
+    set_brightness(new_backlight);
+}
+
+
+//Increases the backlight by as close to the current percentage as surfman allows.
+void decrease_backlight(struct arg_node * args) {
+
+    struct arg_node * node = get_arg(args, 0);
+    unsigned int old_backlight;
+    int new_backlight;
+
+    old_backlight = get_brightness();
+    new_backlight = old_backlight - node->arg.i;
+
+    if (new_backlight < 0) {
+        new_backlight = 0;
+    }
+    else if ((unsigned)new_backlight > old_backlight) {
+        new_backlight = old_backlight;
+    }
+
+    set_brightness((unsigned int)new_backlight);
+}
+
+
+//Get the current backlight in percent.
+static int get_brightness(void) {
+
+    DIR *sys_dir;
+    struct dirent * dp;
+    FILE * file;
+    char data[128];
+    char *device_name = NULL, *path = NULL;
+    int brightness, max_brightness, percent;
+
+    //Get the backlight directory.
+    sys_dir = opendir(BACKLIGHT_PATH);
+    if (!sys_dir) {
+        xcpmd_log(LOG_WARNING, "Couldn't open backlight dir %s - error %d\n", BACKLIGHT_PATH, errno);
+        return 0;
+    }
+
+    //There can be more than one backlight device, and the names of these
+    //devices may vary from platform to platform. They should all report
+    //correct values, so just choose the first one.
+    while ((dp = readdir(sys_dir)) != NULL) {
+        if (dp->d_type == DT_LNK) {
+            device_name = clone_string(dp->d_name);
+            break;
+        }
+    }
+    if (device_name == NULL) {
+        xcpmd_log(LOG_WARNING, "No backlight devices found in %s\n", BACKLIGHT_PATH);
+        closedir(sys_dir);
+        return 0;
+    }
+    closedir(sys_dir);
+
+    //Get the current brightness and max brightness.
+    path = safe_sprintf("%s/%s/%s", BACKLIGHT_PATH, device_name, "actual_brightness");
+    file = fopen(path, "r");
+    if (file == NULL) {
+        xcpmd_log(LOG_WARNING, "Couldn't open file %s - error %d\n", path, errno);
+        free(path);
+        return 0;
+    }
+
+    fgets(data, sizeof(data), file);
+    brightness = atoi(data);
+    fclose(file);
+    free(path);
+
+    path = safe_sprintf("%s/%s/%s", BACKLIGHT_PATH, device_name, "max_brightness");
+    file = fopen(path, "r");
+    if (file == NULL) {
+        xcpmd_log(LOG_WARNING, "Couldn't open file %s - error %d\n", path, errno);
+        free(path);
+        return 0;
+    }
+
+    fgets(data, sizeof(data), file);
+    max_brightness = atoi(data);
+    fclose(file);
+    free(path);
+
+    //Convert backlight to percent.
+    if (max_brightness == 0) {
+        percent = 0;
+    }
+    else {
+        percent = (brightness * 100) / max_brightness;
+    }
+
+    return percent;
+}
+
+
+//Surfman doesn't have a method to set this directly, so we have to do it the
+//roundabout way.
+static void set_brightness(int percent) {
```

Comment:
Lots of duplicated code from the previous function, could probably be refactored.
This time device_name gets free()ed, unless "actual_brightness" or "max_brightness" is not found...